---
title: "Basic-Tutorial"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Basic-Tutorial}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  markdown: 
    wrap: 72
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

Need to install the SeuratData package for the vignette/tutorial, alongside other dependencies. We're
using it just because it has accessible datasets with classifications
(stored in the Seurat metadata field!). Let us know if you have any
issues accessing the data needed for this tutorial.

```{r}
install.packages("seurat")
install.packages("devtools")
install.packages("magrittr")
install.packages("caret")
devtools::install_github("satijalab/seurat-data")
devtools::install_github("mojaveazure/seurat-disk")
devtools::install_github("bnprks/BPCells")
```


```{r}
library(lionmap)
library(SeuratData)
library(Seurat)
library(BPCells)
library(magrittr)
library(caret)
SeuratData::InstallData("pbmc3k")
SeuratObject::UpdateSeuratObject(pbmc3k) #just Seurat updating things to get it to work
```

1.  Get gene expression data in a BPCells matrix and cell type
    classifications for each cell in the dataset.

Steps: a. Set a variable to the gene expression matrix of counts stored
in the Seurat object. These should be raw counts, not processed, scaled,
or normalized post-alignment.

b.  Convert the matrix into a BPCells-formatted object using the
    write_matrix_memory or write_matrix_dir functions.

Write_matrix_dir stores the GE data in a file on your computer, which is
one of the reasons why it's less memory intensive than other file
formats.

If you're not working from a Seurat object, you can do this with any of
the BPCells-accepted file formats (normal matrices or sparse matrices
currently). See the BPCells vignette for more info on how to load your
file into the BPCells format (found here:
<https://bnprks.github.io/BPCells/index.html> ).

c.  Get a dataframe containing metadata with each cell represented as a
    row. We'll store the classifications of each cell as a column in the
    metadata, and the IDs of each cell as another column. Since Seurat
    already provides classifications in their metadata dataframe in the
    seurat_annotations column, we're all set for that column! Seurat
    stores cell IDs as rownames, so we'll add a column called cell_label
    to represent the cell IDs with the same info

```{r}
#a. Set a variable to the gene expression matrix of counts stored in the Seurat object. 
ge_matrix = pbmc3k@assays$RNA@counts

#b. Convert the matrix into a BPCells-formatted object using the write_matrix_memory or write_matrix_dir functions. 
ge_bpcells = write_matrix_memory(ge_matrix)

#c. Get the metadata of each cell stored in the Seurat object.
metadata = pbmc3k@meta.data#$seurat_annotations
metadata$cell_label = rownames(metadata)
```

2.  Split the bpcells object and classifications into a train/test set,
    so we can test how well our classification system works using cells
    from the same dataset.

Note: We're going to get rid of cell classes with a low number of cells
(\< 30) in our dataset because there may not be enough info to
accurately assign these cells a classification.

Steps: a. Remove classes with less than 30 cells from the ge and
annotations dataset. b. Make an 80/20 train/test split, where we assign
a variable to a matrix containing the cells we'll use for our reference
(train set) and the cells we'll use for testing whether our
classification process works (test set).

```{r}
#a. Remove classes with less than 30 cells, and any cell with no cell type classification. For the pbmc3k dataset, the data is stored with rows as gene names and columns as cell IDs. 

#this is important - we need to make sure the metadata celltype label column is formatted properly. To work with the package, you need to remove all spaces from the cell type annotations. You can do this in one line by using the str_replace_all function from the stringr package:
metadata$seurat_annotations = gsub(" ", "", metadata$seurat_annotations, fixed = TRUE)
metadata$seurat_annotations <- metadata$seurat_annotations %>% as.factor()#changing column type back as factor
#refactor now


classifications = metadata$seurat_annotations
classes_to_remove = names(table(classifications))[table(classifications) <= 30] 
#only the platelet class has < 30 cells# identify which cells to remove, and get their positions in the bpcells dataset
cells_to_remove <- which(classifications %in% classes_to_remove)
cells_to_remove <- c(cells_to_remove, which(is.na(classifications))) # remove cells with no classification

ge_bpcells = ge_bpcells[,-cells_to_remove]
#make sure to also remove the corresponding classifications of the cells we just
metadata = metadata[-cells_to_remove,]
metadata$seurat_annotations = droplevels(metadata$seurat_annotations)

#b. Make an 80/20 train/test split, where we assign a variable to a matrix containing
#the cells we'll use for our reference (train set) and the cells we'll use for 
#testing whether our classification process works (test set). 

#we'll use the CreateDataPartition function from the caret package to do this. 
#this function takes in a vector of classes, and a proportion to split each class by,
#and returns the ids to keep in the training set. The remaining ids are the ones we
#keep in the testing set. 
train_ids = caret::createDataPartition(metadata$seurat_annotations, p = .8)
train_ids = train_ids$Resample1#just have this line to get a vector, as opposed to a list
                                #with a vector of positions (which is what the function                                     #normally returns)
test_ids = which(! seq(1,nrow(metadata)) %in% train_ids)
train_ge_bpcells = ge_bpcells[,train_ids]
test_ge_bpcells = ge_bpcells[,test_ids]

#Use the same split to separate the metadata vector into
#a train/test split. This works because the indices of the metadata 
#and of the GE matrix match, so a subset to one would give data corresponding to
#the same cells for the subset to the other
train_metadata = metadata[train_ids,]
test_ge_metadata = metadata[test_ids,]

```

Now that we have a reference dataset (train_ge_pbcells) and a vector of
classifications for each cell in the reference (train_classifications),
we're almost ready to begin finding marker genes that separate our cell
classes! We'll then use these marker genes later on to create models
that can classify cells to specific cell types.

As a way to conveniently access these example datasets without needing to use Seurat to
manually get the pbmc3k dataset, the train ge, test ge, + associated metadata are
automatically installed in the package as train_ex_data_bpcells, test_ex_data_bpcells,
train_ex_metadata, test_ex_metadata. You can get these datasets by typing 
data(name of file), ex: data(train_ex_metadata).

There's one more thing we need to do before finding marker genes, and
that is to create a tree that describes the hierarchy of our cell
classifications. A tree describes the structure of the data (see the
Creating Custom Trees page for more info), and is one of the unique
aspects of this package - the ability to create a structure to embed
info on plausible classifications, and to use that info to make better
classifications. For this basic tutorial, we'll just assume there's no
hierarchical structure, and that celltypes are equally likely to be
assigned to any of the possible classes (aka an equal tree). I've
created a helper function, CreateEqualTree, to do this, which just takes
in a vector of possible cell classes, and creates an equal tree for you.

You'll need to install the treeio package from Bioconductor to get this
to work. Check it out [here for install
info](https://bioconductor.org/packages/release/bioc/html/treeio.html):

```{r}
possible_cell_classes = train_metadata$seurat_annotations %>% unique()
equal_tree = CreateEqualTree(cell_labels = possible_cell_classes)
#OPTIONAL: let's visualize the tree we created (have to install the ggtree package on bioconductor to do this)
#ggtree(equal_tree) +geom_nodelab(geom = "label", node = "all") 
```

Awesome, now we can find marker genes using our reference dataset. The
FindMarkerGenes function does this for us. We'll need to input the
following info:

1.  Our reference GE dataset in BPCells format (**ref_bpcells**)

2.  Our dataframe containing info on each cell (**ref_metadata**)

3.  The tree we created (**tree**)

4.  The name of the column in the metadata giving the cell type
    (**metadata_cluster_column**)

5.  The name of the column in the metadata giving the cell ID
    (**metadata_cell_label_column**)

And optionally we can change the **ngenes** parameter, which allows us
to choose the number of genes we find that distinguishes each pair of
corresponding cells, or the **n_cells_sampled** parameter, which allows
us to choose the number of cells we sample to find differences for each
particular pair of celltypes.

```{r}
marker_genes = FindMarkerGenes(ref_bpcells = train_ge_bpcells, ref_metadata = train_metadata, tree = equal_tree, metadata_cluster_column = "seurat_annotations", metadata_cell_label_column = "cell_label")
```

If it works, we should get a bunch of nonsense printed to the output
talking about writing temporary files on your computer. There's also a
few automated tests that check whether everything is formatted properly,
and they'll throw error messages with some info on what's going wrong if
something is caught. Under the hood, the file is subsetting the GE
matrix to contain 500 cells from two cell classes, writing that data to
a temporary file on your computer, and then using that dataset to find
marker genes that distinguish the classes. It does this for each of the
possible pairs of cell classes you could create.

We'll also get all the marker genes that distinguish each pair of cell
classes! Let's take a look at the output, stored in the marker_genes
variable. Write the following to explore the output and then read on ahead for more info.

```{r}
head(str(marker_genes), max.level = 2)
```

Since Lionmap works on a hierarchy, we have a rootnode (the common node
from which all cell types descend). That's what the list at the highest
level is named for. And at each subsequent level, we have a sublist with
marker genes that distinguish two cell classes (ex: MemoryCD4T vs B). We
see this for all the matchups.

Now we can use this set of marker genes to create our list of models!

Let's use the **GetModels** function to do this.

Just like before, we'll need to input the following info, just with the
addition of the marker gene list we just created:

1.  The marker gene list (**marker_genes**)

2.  Our reference GE dataset in BPCells format (**ref_bpcells**)

3.  Our dataframe containing info on each cell (**ref_metadata**)

4.  The tree we created (**tree**)

5.  The name of the column in the metadata giving the cell type
    (**metadata_cluster_column**)

6.  The name of the column in the metadata giving the cell ID
    (**metadata_cell_label_column**)

And for this function we can optionally change the **n_cells_sampled**
parameter, which is the number of cells sampled from each celltype to
create models that separate each pairwise class (default set to 500),
the **models_to_include** parameter which lets us change the models
we're creating (should only be used by more advanced users), and the
**npcs** parameter, which changes the number of principal components
used when creating models.

```{r}
models <- GetModels(marker_genes = marker_genes, ref_bpcells = train_ge_bpcells, ref_metadata = train_metadata, tree = equal_tree, metadata_cluster_column = "seurat_annotations", metadata_cell_label_column = "cell_label")

```

If everything goes well, the output will indicate the tests checking
proper formatting passed, and will then list the models as they are
being created.

Let's have a look at the structure of the list with the str function (if you're using Rstudio)

```{r}
str(models, max.level = 2)
```

The structure of this object, just like the marker genes, is a list of
lists. Each sublist provides models that can be used to classify a cell
between one of two classes, and we have models that differentiate all
possible classes at each level of the celltype hierarchy (equal_tree).
It also contains information on how to transform data we want to
classify to be in the same PC-space as the models we just created (if
that didn't mean anything to you, don't worry, it's just something about
properly formatting the data). We'll use all these models in the next
step, to classify cells to specific celltypes.

To classify cells to previously defined celltypes, we'll use the
Classify function and will need to enter the following:

1.  Our dataset with raw (not normalized/processed) GE values of cells
    we want to classify (**bpcells_query**)

2.  The list of models produced by GetModels (**models**)

3.  The tree we created (**tree_struc**)

And we can optionally set the parameter prop_max_threshold, which allows
us to set the confidence threshold we need to have in order to classify
a cell.

The output is a vector giving the celltype classifications for each cell
in the query dataset.

```{r}
query_classifications = Classify(bpcells_query = test_ge_bpcells,models = models,tree_struc = equal_tree)
```

Awesome! Now we have a vector of classifications, formatted so the name
of each element is the cell label and the value is the classification of
that cell. Let's view the first few elements.

```{r}
head(query_classifications)
```

We can view the number of classifications to each type using the table
function.

```{r}
table(query_classifications)

```

We'll see that one cell has the classification "Rootnode". This
indicates that there wasn't enough evidence to properly classify the
cell. If we used a custom hierarchy, classification occurs at each level
of the hierarchy, and stops at a given level of the tree if there's not
enough evidence to continue.
