<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="lionmap">
<title>Importing Data from SeuratData and Getting It Into the Proper Format • lionmap</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Importing Data from SeuratData and Getting It Into the Proper Format">
<meta property="og:description" content="lionmap">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">lionmap</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Basic-Tutorial.html">Intro Tutorial</a>
    <a class="dropdown-item" href="../articles/seurat_obj_vignette.html">Importing Data from SeuratData and Getting It Into the Proper Format</a>
    <a class="dropdown-item" href="../articles/tree_vignette.html">Trees! Create Custom Hierarchies to Enhance Classification</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/jonathan-columbiau/lionmap/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Importing Data from SeuratData and Getting It Into the Proper Format</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jonathan-columbiau/lionmap/blob/HEAD/vignettes/seurat_obj_vignette.Rmd" class="external-link"><code>vignettes/seurat_obj_vignette.Rmd</code></a></small>
      <div class="d-none name"><code>seurat_obj_vignette.Rmd</code></div>
    </div>

    
    
<p>The first step is to install dependencies - SeuratData and a package
called devtools to help us install it. SeuratData is a dataset manager
with multiple commonly used datasets stored in the Seurat object format.
Let us know if you have any issues accessing the data needed for this
tutorial.</p>
<p>Install packages</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"satijalab/seurat-data"</span><span class="op">)</span></span></code></pre></div>
<p>Load libraries</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jonathan-columbiau/lionmap" class="external-link">lionmap</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BPCells</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/topepo/caret/" class="external-link">caret</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SeuratData</span><span class="op">)</span></span></code></pre></div>
<p>Install pbmc3k Seurat object</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">SeuratData</span><span class="fu">::</span><span class="fu">InstallData</span><span class="op">(</span><span class="st">"pbmc3k"</span><span class="op">)</span></span>
<span><span class="fu">SeuratObject</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/UpdateSeuratObject.html" class="external-link">UpdateSeuratObject</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span> <span class="co">#just Seurat updating things to get it to be v5</span></span></code></pre></div>
<ol style="list-style-type: decimal">
<li>Get gene expression data in a BPCells matrix and cell type
classifications for each cell in the dataset.</li>
</ol>
<p>Steps:</p>
<ol style="list-style-type: lower-alpha">
<li><p>Set a variable to the gene expression matrix of counts stored in
the Seurat object. These should be raw counts, not processed, scaled, or
normalized post-alignment.</p></li>
<li><p>Convert the matrix into a BPCells-formatted object using the
write_matrix_memory or write_matrix_dir functions.</p></li>
</ol>
<p>Write_matrix_dir stores the GE data in a file on your computer, which
is one of the reasons why it’s less memory intensive than other file
formats.</p>
<p>If you’re not working from a Seurat object, you can do this with any
of the BPCells-accepted file formats (normal matrices or sparse matrices
currently). See the BPCells vignette for more info on how to load your
file into the BPCells format (found here: <a href="https://bnprks.github.io/BPCells/index.html" class="external-link uri">https://bnprks.github.io/BPCells/index.html</a> ).</p>
<ol start="3" style="list-style-type: lower-alpha">
<li>Get a dataframe containing metadata with each cell represented as a
row. We’ll store the classifications of each cell as a column in the
metadata, and the IDs of each cell as another column. Since Seurat
already provides classifications in their metadata dataframe in the
seurat_annotations column, we’re all set for that column! Seurat stores
cell IDs as rownames, so we’ll add a column called cell_label to
represent the cell IDs with the same info</li>
</ol>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#a. Set a variable to the gene expression matrix of counts stored in the Seurat object. </span></span>
<span><span class="va">ge_matrix</span> <span class="op">=</span> <span class="va">pbmc3k</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span></span>
<span></span>
<span><span class="co">#b. Convert the matrix into a BPCells-formatted object using the write_matrix_memory or write_matrix_dir functions. </span></span>
<span><span class="va">ge_bpcells</span> <span class="op">=</span> <span class="fu">write_matrix_memory</span><span class="op">(</span><span class="va">ge_matrix</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#c. Get the metadata of each cell stored in the Seurat object.</span></span>
<span><span class="va">metadata</span> <span class="op">=</span> <span class="va">pbmc3k</span><span class="op">@</span><span class="va">meta.data</span><span class="co">#$seurat_annotations</span></span>
<span><span class="va">metadata</span><span class="op">$</span><span class="va">cell_label</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Split the bpcells object and classifications into a train/test set,
so we can test how well our classification system works using cells from
the same dataset.</li>
</ol>
<p>Note: We’re going to get rid of cell classes with a low number of
cells (&lt; 30) in our dataset because there may not be enough info to
accurately assign these cells a classification.</p>
<p>Steps: a. Remove classes with less than 30 cells from the ge and
annotations dataset.</p>
<ol start="2" style="list-style-type: lower-alpha">
<li>Make an 80/20 train/test split, where we assign a variable to a
matrix containing the cells we’ll use for our reference (train set) and
the cells we’ll use for testing whether our classification process works
(test set).</li>
</ol>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#a. Remove classes with less than 30 cells, and any cell with no cell type classification. For the pbmc3k dataset, the data is stored with rows as gene names and columns as cell IDs. </span></span>
<span></span>
<span><span class="co">#this is important - we need to make sure the metadata celltype label column is formatted properly. To work with the package, you need to remove all spaces from the cell type annotations. You can do this in one line by using the str_replace_all function from the stringr package:</span></span>
<span><span class="va">metadata</span><span class="op">$</span><span class="va">seurat_annotations</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">" "</span>, <span class="st">""</span>, <span class="va">metadata</span><span class="op">$</span><span class="va">seurat_annotations</span>, fixed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">metadata</span><span class="op">$</span><span class="va">seurat_annotations</span> <span class="op">&lt;-</span> <span class="va">metadata</span><span class="op">$</span><span class="va">seurat_annotations</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="op">)</span><span class="co">#changing column type back as factor</span></span>
<span><span class="co">#refactor now</span></span>
<span></span>
<span></span>
<span><span class="va">classifications</span> <span class="op">=</span> <span class="va">metadata</span><span class="op">$</span><span class="va">seurat_annotations</span></span>
<span><span class="va">classes_to_remove</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">classifications</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">classifications</span><span class="op">)</span> <span class="op">&lt;=</span> <span class="fl">30</span><span class="op">]</span> </span>
<span><span class="co">#only the platelet class has &lt; 30 cells# identify which cells to remove, and get their positions in the bpcells dataset</span></span>
<span><span class="va">cells_to_remove</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">classifications</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">classes_to_remove</span><span class="op">)</span></span>
<span><span class="va">cells_to_remove</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">cells_to_remove</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">classifications</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="co"># remove cells with no classification</span></span>
<span></span>
<span><span class="va">ge_bpcells</span> <span class="op">=</span> <span class="va">ge_bpcells</span><span class="op">[</span>,<span class="op">-</span><span class="va">cells_to_remove</span><span class="op">]</span></span>
<span><span class="co">#make sure to also remove the corresponding classifications of the cells we just</span></span>
<span><span class="va">metadata</span> <span class="op">=</span> <span class="va">metadata</span><span class="op">[</span><span class="op">-</span><span class="va">cells_to_remove</span>,<span class="op">]</span></span>
<span><span class="va">metadata</span><span class="op">$</span><span class="va">seurat_annotations</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/droplevels.html" class="external-link">droplevels</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">$</span><span class="va">seurat_annotations</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#b. Make an 80/20 train/test split, where we assign a variable to a matrix containing</span></span>
<span><span class="co">#the cells we'll use for our reference (train set) and the cells we'll use for </span></span>
<span><span class="co">#testing whether our classification process works (test set). </span></span>
<span></span>
<span><span class="co">#we'll use the CreateDataPartition function from the caret package to do this. </span></span>
<span><span class="co">#this function takes in a vector of classes, and a proportion to split each class by,</span></span>
<span><span class="co">#and returns the ids to keep in the training set. The remaining ids are the ones we</span></span>
<span><span class="co">#keep in the testing set. </span></span>
<span><span class="va">train_ids</span> <span class="op">=</span> <span class="fu">caret</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/caret/man/createDataPartition.html" class="external-link">createDataPartition</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">$</span><span class="va">seurat_annotations</span>, p <span class="op">=</span> <span class="fl">.8</span><span class="op">)</span></span>
<span><span class="va">train_ids</span> <span class="op">=</span> <span class="va">train_ids</span><span class="op">$</span><span class="va">Resample1</span><span class="co">#just have this line to get a vector, as opposed to a list</span></span>
<span>                                <span class="co">#with a vector of positions (which is what the function                                     #normally returns)</span></span>
<span><span class="va">test_ids</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="op">!</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">train_ids</span><span class="op">)</span></span>
<span><span class="va">train_ex_data_bpcells</span> <span class="op">=</span> <span class="va">ge_bpcells</span><span class="op">[</span>,<span class="va">train_ids</span><span class="op">]</span></span>
<span><span class="va">test_ex_data_bpcells</span> <span class="op">=</span> <span class="va">ge_bpcells</span><span class="op">[</span>,<span class="va">test_ids</span><span class="op">]</span></span>
<span></span>
<span><span class="co">#Use the same split to separate the metadata vector into</span></span>
<span><span class="co">#a train/test split. This works because the indices of the metadata </span></span>
<span><span class="co">#and of the GE matrix match, so a subset to one would give data corresponding to</span></span>
<span><span class="co">#the same cells for the subset to the other</span></span>
<span><span class="va">train_ex_metadata</span> <span class="op">=</span> <span class="va">metadata</span><span class="op">[</span><span class="va">train_ids</span>,<span class="op">]</span></span>
<span><span class="va">test_ex_metadata</span> <span class="op">=</span> <span class="va">metadata</span><span class="op">[</span><span class="va">test_ids</span>,<span class="op">]</span></span></code></pre></div>
<p>Now that we have a reference dataset (train_ex_data_pbcells) and a
vector of classifications for each cell in the reference
(train_classifications), we’re almost ready to begin finding marker
genes that separate our cell classes! We’ll then use these marker genes
later on to create models that can classify cells to specific cell
types. Check out the intro tutorial for more info on how to do that.</p>
<p>As a way to conveniently access these example datasets without
needing to use Seurat to manually get the pbmc3k dataset on each of the
tutorials, the train ge, test ge, + associated metadata are
automatically installed in the package as train_ex_data_bpcells,
test_ex_data_bpcells, train_ex_metadata, test_ex_metadata. You can get
these datasets by typing data(name of file), ex:
data(train_ex_metadata). We’ll use this in the other vignettees.</p>
  </main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Jonathan Algoo.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
