<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="lionmap">
<title>Trees! Create Custom Hierarchies to Enhance Classification • lionmap</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Trees! Create Custom Hierarchies to Enhance Classification">
<meta property="og:description" content="lionmap">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">lionmap</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Basic-Tutorial.html">Intro Tutorial</a>
    <a class="dropdown-item" href="../articles/tree_vignette.html">Trees! Create Custom Hierarchies to Enhance Classification</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/jonathan-columbiau/lionmap/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Trees! Create Custom Hierarchies to Enhance Classification</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jonathan-columbiau/lionmap/blob/HEAD/vignettes/tree_vignette.Rmd" class="external-link"><code>vignettes/tree_vignette.Rmd</code></a></small>
      <div class="d-none name"><code>tree_vignette.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jonathan-columbiau/lionmap" class="external-link">lionmap</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: plyr</span></span>
<span><span class="co">#&gt; Loading required package: caret</span></span>
<span><span class="co">#&gt; Loading required package: ggplot2</span></span>
<span><span class="co">#&gt; Loading required package: lattice</span></span>
<span><span class="co">#&gt; Loading required package: data.tree</span></span>
<span><span class="co">#&gt; Loading required package: dplyr</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:plyr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     arrange, count, desc, failwith, id, mutate, rename, summarise,</span></span>
<span><span class="co">#&gt;     summarize</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="co">#&gt; Loading required package: e1071</span></span>
<span><span class="co">#&gt; Loading required package: glmnet</span></span>
<span><span class="co">#&gt; Loading required package: Matrix</span></span>
<span><span class="co">#&gt; Loaded glmnet 4.1-8</span></span>
<span><span class="co">#&gt; Loading required package: magrittr</span></span>
<span><span class="co">#&gt; Loading required package: MASS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'MASS'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:dplyr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     select</span></span>
<span><span class="co">#&gt; Loading required package: ranger</span></span>
<span><span class="co">#&gt; Loading required package: Seurat</span></span>
<span><span class="co">#&gt; Loading required package: SeuratObject</span></span>
<span><span class="co">#&gt; Loading required package: sp</span></span>
<span><span class="co">#&gt; 'SeuratObject' was built under R 4.3.0 but the current version is</span></span>
<span><span class="co">#&gt; 4.3.2; it is recomended that you reinstall 'SeuratObject' as the ABI</span></span>
<span><span class="co">#&gt; for R may have changed</span></span>
<span><span class="co">#&gt; 'SeuratObject' was built with package 'Matrix' 1.6.4 but the current</span></span>
<span><span class="co">#&gt; version is 1.6.5; it is recomended that you reinstall 'SeuratObject' as</span></span>
<span><span class="co">#&gt; the ABI for 'Matrix' may have changed</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'SeuratObject'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect</span></span>
<span><span class="co">#&gt; Loading required package: stringr</span></span>
<span><span class="co">#&gt; Loading required package: testthat</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'testthat'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:magrittr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     equals, is_less_than, not</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:dplyr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     matches</span></span>
<span><span class="co">#&gt; Loading required package: treeio</span></span>
<span><span class="co">#&gt; <span style="color: #949494;">treeio v1.26.0 For help: https://yulab-smu.top/treedata-book/</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; </span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; If you use the ggtree package suite in published research, please cite</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; the appropriate paper(s):</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; </span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; LG Wang, TTY Lam, S Xu, Z Dai, L Zhou, T Feng, P Guo, CW Dunn, BR</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; Jones, T Bradley, H Zhu, Y Guan, Y Jiang, G Yu. treeio: an R package</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; for phylogenetic tree input and output with richly annotated and</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; associated data. Molecular Biology and Evolution. 2020, 37(2):599-603.</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; doi: 10.1093/molbev/msz240</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; </span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; Guangchuang Yu.  Data Integration, Manipulation and Visualization of</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; Phylogenetic Trees (1st edition). Chapman and Hall/CRC. 2022,</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; doi:10.1201/9781003279242</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; </span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; G Yu. Data Integration, Manipulation and Visualization of Phylogenetic</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; Trees (1st ed.). Chapman and Hall/CRC. 2022. ISBN: 9781032233574 </span></span></span>
<span><span class="co">#&gt; Loading required package: tidyr</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'tidyr'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:testthat':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     matches</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:magrittr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     extract</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:Matrix':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand, pack, unpack</span></span></code></pre></div>
<p><strong>What is a Hierarchy (aka Tree)?</strong></p>
<p>Note: The terms "phylogeny", "tree", and "hierarchy" are used
interchangeably here.</p>
<p>A hierarchy, also known as a tree, is a type of graph that
illustrates the relationships among a set of entities (represented as
nodes in a tree). Each node (with the exception of the rootnode)
descends from a parent node, and nodes that descend from the same parent
are more related to each other than are entities that descend from
different parent nodes.</p>
<p><strong>Why use a hierarchy?</strong></p>
<p>This article discusses the use of tree-based functionality in
creating and applying classification models. The main idea behind using
a hierarchy is to incorporate prior knowledge about the relationships
between cell types, typically provided in phylogenetic analyses of cell
types in large-scale reference atlases, to enhance classification speed
and accuracy.</p>
<p><strong>What are the advantages of using a hierarchy?</strong></p>
<p>Hierarchies can be used with both pairwise and non-pairwise
classification methods. However, since this package employs pairwise
methods, the focus here is on benefits in a pairwise context, although
some advantages are applicable to both.</p>
<p><strong>Advantage Number 1: Reduced Computational Resources and
Time</strong></p>
<p>One of the primary benefits of using a hierarchy is reducing the
amount of computations you need to perform to get a classification. This
becomes particularly crucial when your reference atlas has many cell
types (ex: dozens or hundreds). In a pairwise context, using a hierarchy
significantly reduces computational time by reducing the number of
cell-cell comparisons that need to be performed to get a classification.
This is because the number of pairwise comparisons you’ll need to
perform at any given level of a hierarchy is proportional to the square
of the number of celltypes at that level.</p>
<p>Number of comparisons = n(n-1) / 2</p>
<p>As an example, let’s take a cell and say it can be classified into
one of twenty possible celltypes. Let’s say we don’t use a hierarchy,
and consider each element as equally likely to occur. To get a
classification we’ll need to perform 20*(20 - 1) / 2 = 190 comparisons.
To generate a picture of what this tree would look like, run the cell
below.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ggtree</span><span class="op">)</span></span>
<span><span class="va">equal_tree</span> <span class="op">=</span> <span class="fu">lionmap</span><span class="fu">::</span><span class="fu"><a href="../reference/CreateEqualTree.html">CreateEqualTree</a></span><span class="op">(</span>cell_labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span>,<span class="st">"n"</span><span class="op">)</span>,rootnode_name <span class="op">=</span> <span class="st">"Root"</span><span class="op">)</span></span>
<span><span class="fu">ggtree</span><span class="op">(</span><span class="va">equal_tree</span><span class="op">)</span> <span class="op">+</span><span class="fu">geom_nodelab</span><span class="op">(</span>geom <span class="op">=</span> <span class="st">"label"</span>, node <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span> </span></code></pre></div>
<p>Alternatively, let’ say we can group the 20 nodes into 4 groups each
with 5 nodes, and these groups can be grouped into two groups each with
2 nodes, and these 2 nodes group together. Let’s create a tree that
looks like this (we’ll get into what the code is doing later), and
visualize it using the ggtree package.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tree_df</span> <span class="op">=</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>            parent <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"p1"</span>,<span class="st">"p1"</span>,<span class="st">"p1"</span>,<span class="st">"p1"</span>,<span class="st">"p1"</span>,</span>
<span>                       <span class="st">"p2"</span>,<span class="st">"p2"</span>,<span class="st">"p2"</span>,<span class="st">"p2"</span>,<span class="st">"p2"</span>,<span class="st">"p3"</span>,<span class="st">"p3"</span>,<span class="st">"p3"</span>,<span class="st">"p3"</span>,<span class="st">"p3"</span>,</span>
<span>                       <span class="st">"p4"</span>,<span class="st">"p4"</span>,<span class="st">"p4"</span>,<span class="st">"p4"</span>,<span class="st">"p4"</span>,<span class="st">"gp1"</span>,<span class="st">"gp1"</span>,<span class="st">"gp2"</span>,</span>
<span>                       <span class="st">"gp2"</span>,<span class="st">"rootnode"</span>,<span class="st">"rootnode"</span>,<span class="st">"rootnode"</span><span class="op">)</span>,</span>
<span>              node <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"n1"</span>,<span class="st">"n2"</span>,<span class="st">"n3"</span>,<span class="st">"n4"</span>,<span class="st">"n5"</span>,</span>
<span>                       <span class="st">"n6"</span>,<span class="st">"n7"</span>,<span class="st">"n8"</span>,<span class="st">"n9"</span>,<span class="st">"n10"</span>,<span class="st">"n11"</span>,<span class="st">"n12"</span>,<span class="st">"n13"</span>,<span class="st">"n14"</span>,</span>
<span>                       <span class="st">"n15"</span>,<span class="st">"n16"</span>,<span class="st">"n17"</span>,<span class="st">"n18"</span>,<span class="st">"n19"</span>,<span class="st">"n20"</span>,<span class="st">"p1"</span>,<span class="st">"p2"</span>,</span>
<span>                       <span class="st">"p3"</span>,<span class="st">"p4"</span>,<span class="st">"gp1"</span>,<span class="st">"gp2"</span>,<span class="st">"rootnode"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">ggtree</span><span class="op">(</span><span class="fu">treeio</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/as.phylo.html" class="external-link">as.phylo</a></span><span class="op">(</span><span class="va">tree_df</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_nodelab</span><span class="op">(</span>geom <span class="op">=</span> <span class="st">"label"</span>, node <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span> </span></code></pre></div>
<p>The total number of comparisons we need to get a classification now
are: 2 at the top level (rootnode -&gt; gp level), 2 for the second
level (gp -&gt; p level), and 5(5-1)/2 = 10 at the final level (p -&gt;
n level) for a grand total of 2 + 2 + 10 = 14 comparisons! A big
difference, huh.</p>
<p><strong>Advantage Number 2: Ability to assess confidence at
progressively more difficult classification levels.</strong></p>
<p><strong>Advantage Number 3: Avoiding Irrelevant Information that
could Throw Off Classification.</strong></p>
<p>Another advantage of using a hierarchy is that it allows for the
avoidance of irrelevant information, such as the gene expression of
marker genes, during the current classification of interest.</p>
  </main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Jonathan Algoo.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
